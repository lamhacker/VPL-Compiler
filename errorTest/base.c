#include "base.h"

float* mallocVec(int num)
{
    float* vec = malloc(sizeof(float) * num + SSE_ALIGN);
	vec = (float *) align(vec, SSE_ALIGN);
	return vec;
}

void freeVec(float* vec)
{
    free(vec);
}

void setValue(float* vec, float val, int num)
{
    for (int i = 0; i < num; i++)
    {
        vec[i] = val;
    }
}

void printVec(char* rep, float* vec, int num)
{
	printf("%s: ", rep);
    for (int i = 0; i < num; i++)
    {   
        printf("%f ", vec[i]);
    }
	printf("\n");
}


