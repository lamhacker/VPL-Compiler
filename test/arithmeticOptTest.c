#include "base.h"
#define NUM (4)

extern void arithOpt1(long, float*);
extern void arithOpt2(long, float*);
extern void arithOpt3(long, float*);
extern void arithOpt4(long, float*);
extern void arithOpt5(long, float*);
extern void arithOpt6(long, float*);

int main(void) {
	
	float* a = mallocVec(NUM);
	
	printf("arithOpt1: \n");
    setValue(a, 1, NUM);
    arithOpt1(NUM, a);
	printVec("a", a, NUM);
	
	printf("arithOpt2: \n");
    setValue(a, 1, NUM);
    arithOpt2(NUM, a);
	printVec("a", a, NUM);
	
	printf("arithOpt3: \n");
    setValue(a, 1, NUM);
    arithOpt3(NUM, a);
	printVec("a", a, NUM);
	
	printf("arithOpt4: \n");
    setValue(a, 1, NUM);
    arithOpt4(NUM, a);
	printVec("a", a, NUM);
	
	printf("arithOpt5: \n");
    setValue(a, 1, NUM);
    arithOpt5(NUM, a);
	printVec("a", a, NUM);
	
	printf("arithOpt6: \n");
    setValue(a, 1, NUM);
    arithOpt6(NUM, a);
	printVec("a", a, NUM);
	
	freeVec(a);
	
	return 0;
}
