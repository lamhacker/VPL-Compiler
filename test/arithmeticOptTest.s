
.text
.global arithOpt1
.type arithOpt1, @function
.p2align 4,,15

arithOpt1:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 5.0 into %rax
    movq $.const5.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end0

.loop_begin0:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin0
.loop_end0:


    popq    %rbx
    leave
    ret

.text
.global arithOpt2
.type arithOpt2, @function
.p2align 4,,15

arithOpt2:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 1.0 into %rax
    movq $.const1.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end1

.loop_begin1:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin1
.loop_end1:


    popq    %rbx
    leave
    ret

.text
.global arithOpt3
.type arithOpt3, @function
.p2align 4,,15

arithOpt3:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 8.0 into %rax
    movq $.const8.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end2

.loop_begin2:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin2
.loop_end2:


    popq    %rbx
    leave
    ret

.text
.global arithOpt4
.type arithOpt4, @function
.p2align 4,,15

arithOpt4:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 0.5 into %rax
    movq $.const0.5, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end3

.loop_begin3:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin3
.loop_end3:


    popq    %rbx
    leave
    ret

.text
.global arithOpt5
.type arithOpt5, @function
.p2align 4,,15

arithOpt5:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 2.0 into %rax
    movq $.const2.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end4

.loop_begin4:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin4
.loop_end4:


    popq    %rbx
    leave
    ret

.text
.global arithOpt6
.type arithOpt6, @function
.p2align 4,,15

arithOpt6:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rsi, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end5

.loop_begin5:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin5
.loop_end5:


    popq    %rbx
    leave
    ret

.data 
.align 16 
.const5.0:
    .float 5.0 
    .float 5.0 
    .float 5.0 
    .float 5.0

.data 
.align 16 
.const1.0:
    .float 1.0 
    .float 1.0 
    .float 1.0 
    .float 1.0

.data 
.align 16 
.const8.0:
    .float 8.0 
    .float 8.0 
    .float 8.0 
    .float 8.0

.data 
.align 16 
.const0.5:
    .float 0.5 
    .float 0.5 
    .float 0.5 
    .float 0.5

.data 
.align 16 
.const2.0:
    .float 2.0 
    .float 2.0 
    .float 2.0 
    .float 2.0

