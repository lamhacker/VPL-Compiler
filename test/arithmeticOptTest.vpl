func arithOpt1(a)
	a = add(3, 2)
end

func arithOpt2(a)
	a = minus(3, 2)
end

func arithOpt3(a)
	a = mult(4, 2)
end

func arithOpt4(a)
	a = div(2, 4)
end

func arithOpt5(a)
	a = min(2, 4)
end

func arithOpt6(a)
	a = div(mult(minus(add(a, 0), 0), 1), 1)
end
