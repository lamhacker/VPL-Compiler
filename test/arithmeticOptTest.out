arithOpt1: 
a: 5.000000 5.000000 5.000000 5.000000 
arithOpt2: 
a: 1.000000 1.000000 1.000000 1.000000 
arithOpt3: 
a: 8.000000 8.000000 8.000000 8.000000 
arithOpt4: 
a: 0.500000 0.500000 0.500000 0.500000 
arithOpt5: 
a: 2.000000 2.000000 2.000000 2.000000 
arithOpt6: 
a: 1.000000 1.000000 1.000000 1.000000 
